src/QsFmt/SPECIFICATION.md

```diff
@@ -0,0 +1,68 @@
+﻿# Q# Formatter Design Specification Document
+
+QsFmt is a source code formatter for Q#.
+
+## Where to get the formatter
+
+The `Microsoft.Quantum.QSharp.Formatter` NuGet package will be distributed with the rest of Q# on NuGet.org.
+As a dotnet tool, it is installed with the command `dotnet tool install Microsoft.Quantum.QSharp.Formatter`.
+
+## How to use the command line formatter
+
+After installing the tool the command `qsfmt` is used to run the formatter.
+There are two commands supported for the tool: 'format' and 'update'.
+The 'format' command runs the tool as a formatter for Q# code, running only those transformation rules that update whitespace and affect indentation. The underlying code will not be changed.
+The 'update' command allows the tool to be used as a means of updating old Q# code, replacing deprecated syntax with newer supported syntax. These transformation rules change the actual code given to them, and may occasionally fail.
+These two commands work separately and can't be run together. If the user wishes to both format and update their code, they would need to run the tool twice, first with 'update' and then with 'format'.
+Both of these commands should be idempotent, meaning that iterated uses of the commands will not continue to have effects on the inputs.
+
+The tool will initially only support the 'update' command, as the formatting functionality is still being worked on. The 'format' command will be unavailable until it is supported.
+
+## Input and Output
+Input files to the formatter will be the last argument(s) to the tool.
+At least one input argument is expected. If no inputs are given, usage information is printed.
+Inputs can be given as one to many paths to directories, files, or any combination. In directories specified, all files with the '.qs' extension found will be taken as input.
+Basic wild cards may be used in the paths.
+The `--recurse` option can be specified to process the input directories recursively for input files. You can specify this for all input directories, not per-directory.
+The `--backup` option can be specified to create backup files for all input files with the original content in them.
+
+The output of the formatter is to overwrite the input files that it processes.
+
+## Rules that are run
+
+The tool will parse the input into a concrete syntax tree using ANTLR. This tree will then have rules applied to it in the form of transformations on the tree.
+The transformations are separated into two groups: transformations run by the 'format' command, and transformations run by the 'update' command.
+Listed here are some of the transformations we intend to use.
+
+### Formatting Transformations
```

Comment:
Will any of these transformations be configurable, perhaps by project file options or a new config file?

Reply:
Currently this is only configurable by selecting which of the two commands to use. The `update` command will run all of the updating rules, and the `format` command will run all of the formatting rules. I believe there was some talk of having these be more configurable in the future, probably through a config file, but this has yet to be designed through, so the current system is not configurable.

Reply:
Closing this thread since the immediate question seems to be answered - further conversations about the functionality in general can be part of the spec review later this months; we will merge a draft spec earlier, and update it based on that review as needed.